Dockerfile

```diff
@@ -55,59 +49,97 @@ RUN if test -n "$FLUTTER_BRANCH" -a "$FLUTTER_BRANCH" != "stable" ; then \
     fi
 
 # Set up Flutter
+# NOTE You will get a warning "Woah! You appear to be trying to run flutter as root."  
+# and this is to be disregarded since this image is never deployed to production.
+# It is not worth setting up a Docker user or mirroring host, but you may run into 
+# conflicts with the local host repo and the Docker one.
 RUN flutter doctor
 RUN flutter --version
 RUN dart pub get
 
+
+
+# ============== NODEJS INTSALL ==============
+# TODO recommended to move to a SHA based install for security reasons
+FROM flutter AS node
+
+RUN curl -sL https://deb.nodesource.com/setup_17.x -o node_setup.sh && \
+      bash node_setup.sh 1> /dev/null
+RUN apt-get update -q && apt-get install -yq --no-install-recommends \
+      nodejs \
+    && rm -rf /var/lib/apt/lists/*
+# Ensure latest NPM, install global Firebase CLI
+RUN npm install -g npm firebase-tools@11.0.1
+
+
+
+# ============== FLUTTER CODE TESTS ==============
+FROM flutter AS tests
+
+COPY ./ ./
+
+ARG FLUTTER_TEST_BRANCH=stable
+ENV FLUTTER_TEST_BRANCH=$FLUTTER_TEST_BRANCH
+
+# Only test the code here, checking links is purely for site deployment
+# NOTE bash scripts will switch the Flutter branch based on $FLUTTER_TEST_BRANCH
+ENTRYPOINT ["tool/test.sh"]
+
+
+
+# ============== DEV / JEKYLL SETUP ==============
+FROM node AS dev
+
+ENV JEKYLL_ENV=development
+RUN gem install bundler
+COPY Gemfile Gemfile.lock ./
+RUN bundle config set force_ruby_platform true
+RUN bundle install
+
+# Install Node deps
+ENV NODE_ENV=development
+COPY package.json package-lock.json ./
+RUN npm install
+
+COPY ./ ./
+
+# So that we have the most up to date submodules
+RUN git submodule update --init --recursive
+
+# Jekyl ports
 EXPOSE 35729
 EXPOSE 4002
 
+# Firebase emulator port
+# Airplay runs on :5000 by default now
+EXPOSE 5500
 
-# -- Test target
-FROM dev as test
-ARG DISABLE_TESTS
-ENV DISABLE_TESTS=$DISABLE_TESTS
-ARG TEST_TARGET_CHANNEL=stable
-ENV TEST_TARGET_CHANNEL=$TEST_TARGET_CHANNEL
-RUN tool/test.sh --target $TEST_TARGET_CHANNEL --check-links --null-safety
 
 
-# -- Build target
-FROM test AS builder
+# ============== BUILD PROD JEKYLL SITE ==============
+FROM node AS build
 
 ENV JEKYLL_ENV=production
+RUN gem install bundler
+COPY Gemfile Gemfile.lock ./
+RUN bundle config set force_ruby_platform true
+RUN BUNDLE_WITHOUT="test development" bundle install --jobs=4 --retry=2 --quiet
```

Comment:
could we potentially want logs here? Is bundle install too noisy which is why --quiet is being used here?

Reply:
Yes, for the production build. Makes sense, I'll remove that with this commit.